cairo_png.c

```diff
@@ -0,0 +1,135 @@
+#include <assert.h>
+#include <cairo.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <png.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "cairo_png.h"
+
+cairo_status_t cairo_surface_write_to_png_mem(cairo_surface_t *sfc,
+                                              unsigned char **data,
+                                              unsigned long *len) {
+  int png_color_type;
+  int bpc;
+#ifdef PNG_pHYs_SUPPORTED
+  double dpix, dpiy;
+#endif
+
+  int height = cairo_image_surface_get_height(sfc);
+  int width = cairo_image_surface_get_width(sfc);
+  cairo_format_t cformat = cairo_image_surface_get_format(sfc);
+
+  /* PNG complains about "Image width or height is zero in IHDR" */
+  if (width == 0 || height == 0) {
+    return CAIRO_STATUS_WRITE_ERROR;
+  }
+
+  png_structp png =
+    png_create_write_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+
+  png_infop info = png_create_info_struct(png);
+  if (info == NULL) {
+    png_destroy_write_struct(&png, &info);
+    return CAIRO_STATUS_NO_MEMORY;
+  }
+
+#ifdef PNG_SETJMP_SUPPORTED
+  if (setjmp(png_jmpbuf(png))) {
+    png_destroy_write_struct(&png, &info);
+    return 2;
+  }
+#endif
+
+  //png_set_write_fn(png, &state, my_png_write_data, png_simple_output_flush_fn);
+
+  FILE *f = fopen("/tmp/meow.png", "wb");
+  if (f == NULL) {
+    fprintf(stderr, "failed to open output file\n");
+    exit(EXIT_FAILURE);
+  }
+  png_init_io(png, f);
+
+  switch (cformat) {
+  case CAIRO_FORMAT_ARGB32:
+    bpc = 8;
+    png_color_type = PNG_COLOR_TYPE_RGB_ALPHA;
+    break;
+  case CAIRO_FORMAT_INVALID:
+  default:
+    png_destroy_write_struct(&png, &info);
+    return CAIRO_STATUS_INVALID_FORMAT;
+  }
+
+  png_set_IHDR(png, info, width, height, bpc, png_color_type,
+               PNG_INTERLACE_NONE, PNG_COMPRESSION_TYPE_DEFAULT,
+               PNG_FILTER_TYPE_DEFAULT);
+
+#ifdef PNG_pHYs_SUPPORTED
+  cairo_surface_get_fallback_resolution(sfc, &dpix, &dpiy);
```

Comment:
We should get the px/mm value from the monitor's physical size: https://github.com/emersion/grim/blob/master/main.c#L113

Reply:
Righto. For the case of a single output it would just be something like pixels_x/physical_width, but now I'm wondering how this should work for when the screenshot contains multiple outputs.
Just sum them all together?

Reply:
Yeah, except the screenshot region could partially span over multiple outputs. In general it may be even more complicated because different outputs could cover the same region.

I wonder if there's a good way to handle this.